### PR TITLE
Fixing hyperblock partial mb handling part2

### PR DESCRIPTION
**Pull request by iulianpascalau**
## Reasoning behind the pull request
- added blockReverted functionality + tests
  
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: dblookupext/miniblocksHandler.go [hunk 1]
```diff
@@ -197,3 +197,88 @@ func (mh *miniblocksHandler) saveMiniblocksMetadata(miniblockHash []byte, minibl
...
+
+func (mh *miniblocksHandler) revertMiniblock(header data.HeaderHandler, headerHash []byte, mb *block.MiniBlock, miniblockHash []byte) error {
+	miniblockHeader, err := mh.findMiniblockHandler(header, miniblockHash)
+	if err != nil {
+		return err
+	}
+
+	if miniblockHeader.GetIndexOfFirstTxProcessed() == 0 {
+		// we either revert a complete miniblock or the first partial executed miniblock header
+		err = mh.epochIndex.removeEpochByHash(miniblockHash)
```
Comment (andreibancioiu) [inline dblookupext/miniblocksHandler.go:230]: Nice to spot this cleanup case.

File: dblookupext/miniblocksHandler.go [hunk 2]
```diff
@@ -197,3 +197,88 @@ func (mh *miniblocksHandler) saveMiniblocksMetadata(miniblockHash []byte, minibl
...
+		if !bytes.Equal(mbOnBlock.HeaderHash, headerHash) {
+			newMiniblocksInfo = append(newMiniblocksInfo, mbOnBlock)
+		}
+	}
+	loadedMiniblockMetadata.MiniblocksInfo = newMiniblocksInfo
+
+	if len(loadedMiniblockMetadata.MiniblocksInfo) == 0 {
+		// we assume that the rollback is done in-sync with the node.
+		// TODO: refactor pruningStorer and add RemoveFromEpoch functionality
+		err = mh.miniblocksMetadataStorer.RemoveFromCurrentEpoch(miniblockHash)
```
Comment (andreibancioiu) [inline dblookupext/miniblocksHandler.go:278]: Can this be moved after L266 (optional)?

File: dblookupext/miniblocksHandler.go [hunk 3]
```diff
@@ -197,3 +197,88 @@ func (mh *miniblocksHandler) saveMiniblocksMetadata(miniblockHash []byte, minibl
...
+	for _, mbOnBlock := range loadedMiniblockMetadata.MiniblocksInfo {
+		if !bytes.Equal(mbOnBlock.HeaderHash, headerHash) {
+			newMiniblocksInfo = append(newMiniblocksInfo, mbOnBlock)
+		}
+	}
+	loadedMiniblockMetadata.MiniblocksInfo = newMiniblocksInfo
+
+	if len(loadedMiniblockMetadata.MiniblocksInfo) == 0 {
+		// we assume that the rollback is done in-sync with the node.
+		// TODO: refactor pruningStorer and add RemoveFromEpoch functionality
```
Comment (andreibancioiu) [inline dblookupext/miniblocksHandler.go:277]: :+1:

File: dblookupext/miniblocksHandler.go [hunk 4]
```diff
@@ -197,3 +197,88 @@ func (mh *miniblocksHandler) saveMiniblocksMetadata(miniblockHash []byte, minibl
...
+	newMiniblocksInfo := make([]*MiniblockMetadataOnBlock, 0, len(loadedMiniblockMetadata.MiniblocksInfo))
+	for _, mbOnBlock := range loadedMiniblockMetadata.MiniblocksInfo {
+		if !bytes.Equal(mbOnBlock.HeaderHash, headerHash) {
+			newMiniblocksInfo = append(newMiniblocksInfo, mbOnBlock)
+		}
+	}
+	loadedMiniblockMetadata.MiniblocksInfo = newMiniblocksInfo
+
+	if len(loadedMiniblockMetadata.MiniblocksInfo) == 0 {
+		// we assume that the rollback is done in-sync with the node.
```
Comment (andreibancioiu) [inline dblookupext/miniblocksHandler.go:276]: We always revert something from the current epoch to something that might be from the current one or previous one - correct?

**Review by andreibancioiu — APPROVED** [review]

File: dblookupext/miniblocksHandler.go (hunk 1 above)
Comment (iulianpascalau) [reply dblookupext/miniblocksHandler.go:230]: right, this is how the miniblock tracker is implemented in the node

File: dblookupext/miniblocksHandler.go (hunk 2 above)
Comment (iulianpascalau) [reply dblookupext/miniblocksHandler.go:278]: Should work but will stress the storer. I would keep it like this but I can refactor the code to look better.

File: dblookupext/miniblocksHandler.go (hunk 4 above)
Comment (iulianpascalau) [reply dblookupext/miniblocksHandler.go:276]: absolutely. The process is like this:
`revert block <epoch e, nonce n>` 
`revert block <epoch e, nonce n-1>`
`change epoch to e-1`
`revert block <epoch e-1, nonce n-2>`
`revert block <epoch e-1, nonce n-3>`
....
the revert is always done in-sync

**Review by andreibancioiu — APPROVED** [review]

File: dblookupext/miniblocksHandler_test.go [hunk 5]
```diff
@@ -493,3 +493,204 @@ func TestMiniblocksHandler_loadExistingMiniblocksMetadataWithLegacy(t *testing.T
...
+	assert.Nil(t, err)
+	err = checkTxs(mbHandler.miniblockHashByTxHashIndexStorer, header1.GetEpoch(), partialMiniblock.TxHashes[0:4], partialMiniblockHash)
+	assert.Nil(t, err)
+	err = checkTxs(mbHandler.miniblockHashByTxHashIndexStorer, header3.GetEpoch(), partialMiniblock.TxHashes[5:6], partialMiniblockHash)
+	assert.NotNil(t, err) // these were reverted
+
+	// check the miniblock hashes were written correctly
+	err = checkMiniblockHashInEpoch(mbHandler.epochIndex, completeMiniblockHash, header1.Epoch)
+	assert.Nil(t, err)
+	// partial miniblock was first notarized in epoch 37
```
Comment (bogdan-rosianu) [inline dblookupext/miniblocksHandler_test.go:557]: :+1:

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**